Improve usability of zooming in the animation editor

CookieBadger opened the pull request:
This PR seeks to improve the zoom usability in the AnimationPlayer editor, I would be happy to hear your feedback! Especially the changes I had to make to the header file is something that a more experienced pair of eyes should be laid on.

I added some code in the AnimationTrackEditor, such that zooming with the mouse zooms in on the current mouse position, whereas the zoom panner always zooms in on the center of the currently visible timeline, instead of zooming on the beginning in both cases.
This implementation comes as a result of the proposal [#6204](https://github.com/godotengine/godot-proposals/issues/6204)

I have never contributed before, so please bear with me :)

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

CookieBadger committed:
Improved usability of zoom features in animation player

CookieBadger commented:
Instead of always zooming in on the left side of the timeline, this is the way zooming behaves now:
![AnimationPlayerZoom](https://github.com/godotengine/godot/assets/53947784/4c509964-d680-492d-9d18-c309e6558b8c)

Also works in the Bezier editor:
![AnimationBezierZoom](https://github.com/godotengine/godot/assets/53947784/973b147a-9ad5-4ab2-bf86-a6919d8ba821)

CookieBadger commented on editor/animation_track_editor.h:
```diff
@@ -184,6 +188,7 @@ class AnimationTimelineEdit : public Range {
 	void set_track_edit(AnimationTrackEdit *p_track_edit);
 	void set_zoom(Range *p_zoom);
 	Range *get_zoom() const { return zoom; }
+	void _zoom_callback(float p_zoom_factor, Vector2 p_origin, Ref<InputEvent> p_event);
```
Making the callback public is probably not a good idea, should I instead wrap it in a separate method like `on_zoom()` that is called in `_zoom_callback()`, that the outer classes then call?

AThousandShips replied:
Either that or use `friend` for the classes that need it, the callback should remain private

CookieBadger replied:
good point, thanks!

fire commented:
I like it! @TokageItLab What do you think?

Calinou reviewed (commented):
Tested locally, there are some issues:

- The zoom is slightly offset from the mouse cursor position (or the center when using the zoom bar). This seems to vary depending on the key column's size and the editor window width.
- Editor scales above 100% further offset the zoom (see the second video). Perhaps some multiplying/dividing by `EDSCALE` needs to be done somewhere.

This has the potential to be a big productivity boost, so it's encouraging to see this being worked on nonetheless :slightly_smiling_face: 

### 100% editor scale

https://github.com/godotengine/godot/assets/180032/49230ffd-cbec-4f86-b15c-c58b79ddb16c

### 200% editor scale

https://github.com/godotengine/godot/assets/180032/681dae9c-062d-4a2f-bc3e-18cfafe955a6

TokageItLab commented:
I think it is a good way for improvement. I will test it as soon as some of the points already pointed out are fixed.

CookieBadger commented:
Thank you for pointing the issues out, the zoomed timeline length was not calculated correctly. Should be fixed now!

Calinou reviewed (approved):
Tested locally (rebased on top of `master` d76c1d0e516fedc535a2e394ab780cac79203477), it works as expected (tested at 100% and 150% scale factors).

TokageItLab reviewed (changes requested):
When using the mouse wheel, it makes sense and works correctly based on the current cursor position. Nice work.

However, when using the slider in the lower right corner, the zoom should be based on the current position (e.g. Affter Effects bahave that). Especially when changing the zoom factor from small to large, I believe it is an odd behavior to have the end of the animation as the reference.

Current this PR:

https://github.com/godotengine/godot/assets/61938263/7bd825f4-0aa2-498b-9795-d96788092fd3

FYI Affter Effects:

https://github.com/godotengine/godot/assets/61938263/5d0adb05-4ca5-4452-a2f6-0fbb0985ae4e

CookieBadger commented:
Thanks for the suggestion @TokageItLab , makes sense. Something that I don't like about AfterEffects' approach though is that the timeline view switches position such that the cursor is centered upon zooming (0:06 in the video). I think this is quite distracting/disturbing, and it would be more appropriate if the zoom focused on the cursor, but didn't make the timeline jump around, like this:
![AnimationPlayerZoomFocusOnPosition](https://github.com/godotengine/godot/assets/53947784/269e6532-ff1a-4bbd-95ec-e40eb4add5c4)

However then, an edge case that makes that approach problematic, is what to do if the cursor position is outside the currently displayed range. It might again be distracting or disturbing if it automatically switches to the middle, but with the current approach it is unclear what the zoom is actually focusing on.
![AnimationPlayerZoomFocusOnPositionOutside](https://github.com/godotengine/godot/assets/53947784/a44b3358-7008-4419-b348-20172f77e46f)

Alternatively, I could test for this case, and then just zoom on the middle of the timeline as before, but I'd like to know what you think.

TokageItLab commented:
BTW, the first video in your reply is a little different from the behavior of AffterEffects: it should zooms based on the right or left edge of the window until the current position is centered. Then, after the current position is centered, it zooms based on the current position. The jump in window position is done when the window is shrunk (I also feel it is a bit odd in AffterEffects).

If the current position is not on the window, it is a matter of preference.

If the current position is not on the window and is based on the center of the screen as Yuri suggests, you may need to switch the behavior after the current position appears on the screen.

Personally, I don't care much about the timeline view switching positions, so if it is easiest to change the timeline view position, that is fine.

CookieBadger commented:
I now implemented the zooming similar to your suggestion, the way AfterEffects does it, such that the timeline stretches from the sides until the play position is in the middle.

https://github.com/godotengine/godot/assets/53947784/a37d4289-8688-4fae-b442-3f683902f56e

In the edge case of the cursor being outside, it now zooms into the middle, or zooms out from the middle until the play position comes into view, then again zooms out on the cursor

https://github.com/godotengine/godot/assets/53947784/50326234-8c5e-4449-846a-c0401758588c

I hope every case is properly addressed now.

CookieBadger commented:
@TokageItLab would you like to see any other changes?

TokageItLab reviewed (approved):
Now it seems to be working fine overall.

One concern is that the center appears to shake when zooming with the slider, so it would be nice if this could be removed, but it can be followed up later as another PR.

https://github.com/godotengine/godot/assets/61938263/1d46b00d-4e71-4308-b2e3-337a6d0efcd2

AThousandShips commented on editor/animation_track_editor.cpp:
```diff
@@ -1239,6 +1239,56 @@ void AnimationMultiTrackKeyEdit::set_use_fps(bool p_enable) {
 }
 
 void AnimationTimelineEdit::_zoom_changed(double) {
+	double zoom_pivot = 0; // point on timeline to stay fixed
```
```suggestion
	double zoom_pivot = 0; // Point on timeline to stay fixed.
```
Per the [comment style](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide). Same for all your code

CookieBadger commented:
Fixed comment style, added #define constant, and added a tiny offset to prevent the wiggling that @TokageItLab mentioned. I think the PR is ready for merge now.

https://github.com/godotengine/godot/assets/53947784/3ed16940-a9c6-4ed8-a512-2cee2199d0b9

TokageItLab commented on editor/animation_track_editor.cpp:
```diff
@@ -6479,6 +6539,7 @@ AnimationTrackEditor::AnimationTrackEditor() {
 	timeline->connect("length_changed", callable_mp(this, &AnimationTrackEditor::_update_length));
 
 	panner.instantiate();
+	panner->set_scroll_zoom_factor(1.02);
```
```suggestion
	panner->set_scroll_zoom_factor(SCROLL_ZOOM_FACTOR);
```
Replace 1.02 in BezierEditor as well.

CookieBadger replied:
woops, I missed that one, sorry. Resolved.

CookieBadger commented:
any other requests? I'd be happy to finish this by the end of the year 😄

KoBeWi commented on editor/animation_track_editor.h:
```diff
@@ -41,6 +41,8 @@
 #include "scene/gui/tree.h"
 #include "scene/resources/animation.h"
 
+#define SCROLL_ZOOM_FACTOR 1.02 // Zoom factor per mouse scroll in the animation editor. The closer to 1.0, the finer the control.
```
This should be defined as `static constexpr float`, preferably inside the class.

CookieBadger replied:
moved it down to AnimationTimelineEdit, I think that's where it belongs the most

KoBeWi commented on editor/animation_track_editor.h:
```diff
@@ -128,6 +130,8 @@ class AnimationMultiTrackKeyEdit : public Object {
 
 class AnimationTimelineEdit : public Range {
 	GDCLASS(AnimationTimelineEdit, Range);
+	friend class AnimationBezierTrackEdit;
+	friend class AnimationTrackEditor;
```
```suggestion
	GDCLASS(AnimationTimelineEdit, Range);

	friend class AnimationBezierTrackEdit;
	friend class AnimationTrackEditor;
```

KoBeWi commented on editor/animation_track_editor.cpp:
```diff
@@ -1239,6 +1239,59 @@ void AnimationMultiTrackKeyEdit::set_use_fps(bool p_enable) {
 }
 
 void AnimationTimelineEdit::_zoom_changed(double) {
```
Shouldn't the variables here be `float`? Most methods called here use float.

CookieBadger replied:
I am not sure. `get_value()` returns a double, and `hscroll` (derives `Range`) expects a double, so I thought I'd just do all calculations with doubles. Should I change everything to be a float instead?

KoBeWi commented on editor/animation_track_editor.cpp:
```diff
@@ -1239,6 +1239,59 @@ void AnimationMultiTrackKeyEdit::set_use_fps(bool p_enable) {
+	double updated_timeline_half_width = updated_timeline_width_seconds / 2.0;
+	bool zooming = updated_timeline_width_seconds < timeline_width_seconds;
+
+	double timeline_left = get_value();
+	double timeline_right = get_value() + timeline_width_seconds;
+	double timeline_center = get_value() + timeline_width_seconds / 2.0;
```
```suggestion
	double timeline_left = get_value();
	double timeline_right = timeline_left + timeline_width_seconds;
	double timeline_center = timeline_left + timeline_width_seconds / 2.0;
```

KoBeWi commented on editor/animation_track_editor.cpp:
```diff
@@ -1239,6 +1239,59 @@ void AnimationMultiTrackKeyEdit::set_use_fps(bool p_enable) {
+		zoom_pivot = timeline_left + timeline_width_seconds * zoom_scroll_origin_norm;
+		zoom_pivot_delta = updated_timeline_width_seconds * zoom_scroll_origin_norm;
+		zoom_callback_occured = false;
+	} else { // Zooming with panner will depend on the current play position.
```
```suggestion
	} else { // Zooming with slider will depend on the current play position.
```
Typo?

KoBeWi commented on editor/animation_track_editor.cpp:
```diff
@@ -1239,6 +1239,59 @@ void AnimationMultiTrackKeyEdit::set_use_fps(bool p_enable) {
+		zoom_pivot_delta = updated_timeline_width_seconds * zoom_scroll_origin_norm;
+		zoom_callback_occured = false;
+	} else { // Zooming with panner will depend on the current play position.
+		const double epsilon = 0.00001;
```
You can use `CMP_EPSILON`.

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:

CookieBadger commented:
Thank you!! 😊
